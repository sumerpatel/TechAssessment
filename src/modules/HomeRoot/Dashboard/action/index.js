const namespacer = (namespace) => (action) => {
  return `${namespace}/${action}`;
};

const namespacedAction = namespacer('DashboardReducer');

export const actionTypes = {
  GET_DETAILS: namespacedAction('GET_DETAILS'),
  GET_DETAILS_RESPONSE: namespacedAction('GET_DETAILS_RESPONSE'),
  GET_DETAILS_ERROR: namespacedAction('GET_DETAILS_ERROR'),
};

export const actions = {
  getUserData() {
    return {
      type: actionTypes.GET_DETAILS,
      payload: {},
    };
  },
  updateUserDataResponse(data) {
    return {
      type: actionTypes.GET_DETAILS_RESPONSE,
      data: data,
    };
  },
  updateUserDataError(error) {
    return {
      type: actionTypes.GET_DETAILS_ERROR,
      error: error,
    };
  },
};
