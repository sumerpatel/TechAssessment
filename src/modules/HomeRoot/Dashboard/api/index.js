import CommonService from '../../../../service';

export default class DashboardApi extends CommonService {
  async getUserDataCall() {
    try {
      const url = 'https://itunes.apple.com/search?term=jack+johnson';
      return this.get(url);
    } catch (error) {
      throw error;
    }
  }
}

export const dashboardService = new DashboardApi();
