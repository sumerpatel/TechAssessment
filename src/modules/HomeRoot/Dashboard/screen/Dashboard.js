import React from 'react';
import {
  FlatList,
  Text,
  View,
  ActivityIndicator,
  Image,
  TouchableOpacity,
} from 'react-native';
import { AppColors } from '../../../../common/Colors';
import { navigate } from '../../../../navigation/NavigationService';
import styles from './styles';

class Dashboard extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      userData: [],
    };
  }

  componentDidMount() {
    this.props.getUserData();
  }

  componentDidUpdate(prevProps) {
    const { userData } = this.props;
    if (userData && prevProps.userData !== userData) {
      this.setState({ userData: userData.results });
    }
  }

  renderLoader = () => {
    return <ActivityIndicator size="large" color={AppColors.PRIMARY} />;
  };

  renderItem = (item, index) => {
    return (
      <TouchableOpacity
        onPress={() => navigate('TrackDetails', { data: item })}
        style={styles.itemContainer}>
        <Image
          style={{ height: 100, width: 100 }}
          source={{ uri: item.artworkUrl100 }}
        />
        <Text style={[styles.textStyle, { fontWeight: 'bold' }]}>
          {item.trackName}
        </Text>
      </TouchableOpacity>
    );
  };

  render() {
    const { userData } = this.state;
    return (
      <View style={styles.container}>
        {userData && userData.length > 0 ? (
          <FlatList
            data={userData}
            renderItem={({ item, index }) => this.renderItem(item, index)}
            keyExtractor={(item, index) => index.toString()}
            numColumns={2}
          />
        ) : (
          this.renderLoader()
        )}
      </View>
    );
  }
}
export default Dashboard;
