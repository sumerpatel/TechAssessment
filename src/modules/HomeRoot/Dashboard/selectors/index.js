import { get } from 'lodash';

export const reducer = (state) => {
  return get(state, 'DashboardReducer', {});
};

export const getDetails = (state) => {
  const data = reducer(state);
  return get(data, 'userData', undefined);
};
