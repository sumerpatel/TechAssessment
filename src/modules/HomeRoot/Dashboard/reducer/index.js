import { actionTypes } from '../action';

const initialState = {
    loading: false,
    userData: undefined,
};

export const DashboardReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.GET_DETAILS:
      return {
        ...state,
        loading: true
      };
    case actionTypes.GET_DETAILS_RESPONSE:
      return {
        ...state,
        loading: false,
        userData: action.data,
      };
    case actionTypes.GET_DETAILS_ERROR:
      return {
        ...state,
        loading: false,
        userData: action.error,
      };

    default:
      return state;
  }
};
export default DashboardReducer;
