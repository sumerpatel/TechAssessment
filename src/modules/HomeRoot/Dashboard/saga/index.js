/**
 * @module Sagas
 * @module Sagas/Bill
 * @desc User
 */
import { all, put, takeLatest } from 'redux-saga/effects';
import { actions, actionTypes } from '../action';
import { dashboardService } from '../api';

export function* getUserData() {
  try {
    const response = yield dashboardService.getUserDataCall();
    yield put(actions.updateUserDataResponse(response));
  } catch (err) {
    yield put(actions.updateUserDataError(err));
  }
}

/**
 * Root Sagas
 */
export default function* dashboardRoot() {
  yield all([takeLatest(actionTypes.GET_DETAILS, getUserData)]);
}
