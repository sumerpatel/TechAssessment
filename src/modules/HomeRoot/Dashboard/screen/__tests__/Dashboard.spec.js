import React from 'react';
import { shallow } from 'enzyme';
import Dashboard from '../Dashboard';

describe('Dashboard view render', () => {
  let wrapper;
  const props = {
    userData: undefined,
    getUserData: jest.fn(),
  };
  beforeEach(() => {
    wrapper = shallow(<Dashboard {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
