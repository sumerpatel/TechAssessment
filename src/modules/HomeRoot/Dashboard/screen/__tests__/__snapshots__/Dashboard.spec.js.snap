// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard view render should render correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#E30613"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
